pkg/utils/object.go

```diff
@@ -0,0 +1,224 @@
+// The full content of the object is returned as map[string]interface{}, except for system metadata fields.
+// This function can be combined with runtime.DefaultUnstructuredConverter.FromUnstructured to get the object content
+// as runtime.RawExtension.
+func GetObject(ctx context.Context, c client.Client, gvk schema.GroupVersionKind, name, namespace string) (map[string]interface{}, error) {
+	// Initialize the object
+	key := client.ObjectKey{Namespace: namespace, Name: name}
+	obj := &unstructured.Unstructured{}
+	obj.SetGroupVersionKind(gvk)
+
+	// Get the object
+	if err := c.Get(ctx, key, obj); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(err, "could not get object %s %s", gvk, key)
+		}
+
+		// The object was not found
+		return nil, nil
+	}
+
+	return filterMetadata(obj.UnstructuredContent(), systemMetadataFields...), nil
+}
+
+// CreateOrUpdateObjectByRef creates or updates the object with the given reference and namespace using the given client.
+// The object is created or updated with the given content, except for system metadata fields.
+// This function can be combined with runtime.DefaultUnstructuredConverter.ToUnstructured to create or update an object
+// from runtime.RawExtension.
+func CreateOrUpdateObjectByRef(ctx context.Context, c client.Client, ref *autoscalingv1.CrossVersionObjectReference, namespace string, content map[string]interface{}) error {
+	gvk, err := gvkFromCrossVersionObjectReference(ref)
+	if err != nil {
+		return err
+	}
+	return CreateOrUpdateObject(ctx, c, gvk, ref.Name, namespace, content)
+}
+
+// CreateOrUpdateObject creates or updates the object with the given GVK, name, and namespace using the given client.
+// The object is created or updated with the given content, except for system metadata fields, namespace, and name.
+// This function can be combined with runtime.DefaultUnstructuredConverter.ToUnstructured to create or update an object
+// from runtime.RawExtension.
+func CreateOrUpdateObject(ctx context.Context, c client.Client, gvk schema.GroupVersionKind, name, namespace string, content map[string]interface{}) error {
+	// Initialize the object
+	key := client.ObjectKey{Namespace: namespace, Name: name}
+	obj := &unstructured.Unstructured{}
+	obj.SetGroupVersionKind(gvk)
+	obj.SetName(name)
+	obj.SetNamespace(namespace)
+
+	// Create or update the object with retries for optimistic concurrency
+	for retries, done := 0, false; !done; retries++ {
+		// Check if the object already exists
+		found := true
+		if err := c.Get(ctx, key, obj); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return errors.Wrapf(err, "could not get object %s %s", gvk, key)
+			}
+
+			// The object was not found
+			found = false
+		}
+
+		// Set object content
+		if content != nil {
+			obj.SetUnstructuredContent(mergeObjectContents(obj.UnstructuredContent(),
+				filterMetadata(content, add(systemMetadataFields, "namespace", "name")...)))
+		}
+
+		// Create or update the object
```

Comment:
Why not calling `controllerutil.CreateOrUpdate`?

Reply:
Done